feature: address typescript issues

murbanowicz opened the pull request:
This PR moves whole codebase to TypeScript and enables build for ESM and UMD, including the types.

I tested it quickly by installing the package with 1. npm pack 2. adding it to package json with 'file:/...' and it works nicely:
![image](https://user-images.githubusercontent.com/9385106/206465625-31bc603f-dbd4-45d7-a35e-5db671fd971e.png)

It depends on types from kinde-pkce-js which should be adjusted and made more precise, but then the types in this package will auto update after updating this dependency as they rely on the types of the KindeClient exported from that library.

You may want to adjust it, but I believe majority of work is done.

**TERSER is not added to the build**. It was breaking the build completely to me and I had to remove it at this stage. 
I am not that familiar with Rollup to get it working in reasonable timeframe.
Long story short - after adding terser, it starts to break with weird CJS error on config, while without it, it works fine.

Feel free to contact me on Slack if you need any more help with this and obviously to edit this by adding other required stuff.

DaveOrDead commented on rollup.config.js:
```diff
@@ -1,35 +1,41 @@
+const commonPlugins = [
+  typescript({ tsconfig: './tsconfig.json' }),
+  commonjs(),
+    // terser()
```
Hey @murbanowicz thanks for the PR - looking awesome 🚀 

Just wondering if there was a reason for removing minification?

murbanowicz replied:
Hi Mate! Look at the bottom part of the PR description. I would ask you to help on this and manage rollup. I've spent quite a lot of time on this already which makes it pointless on my side.

DaveOrDead replied:
@murbanowicz my bad, missed that part of the description. Looking into it now :)

We've updated the types in the PKCE vanilla JS lib this depends on now also which may make some of the typings here better

DaveOrDead replied:
It seems that this issue is with upgrading terser from `0.1.0` to `0.2.0` as per [this issue](https://github.com/rollup/plugins/issues/1366) 

If you revert the change in `package.json` back to `"@rollup/plugin-terser": "^0.1.0"` then it all builds fine :) 

Hopefully they resolve soon so we can use the latest, but I'm ok with it being on 0.1.0 for now

DaveOrDead commented on src/state/reducer.ts:
```diff
@@ -1,27 +1,30 @@
         ...state,
         isAuthenticated: !!action.user,
         user: action.user
       };
```
I've realised `onComplete` actually isn't used anywhere so this whole function can be removed

DaveOrDead commented on src/state/reducer.ts:
```diff
@@ -35,10 +38,8 @@ const reducerMap = {
   ERROR: onError
```
As we're removing the `onComplete` function, these two lines can also be removed
```
ON_REDIRECT_COMPLETE: onComplete,
GET_ACCESS_TOKEN_COMPLETE: onComplete,
```

DaveOrDead commented on src/state/types.ts:
```diff
@@ -0,0 +1,14 @@
+  id: string | null;
+  family_name: string | null;
+  email: string | null;
+  updated_at: number;
```
`updated_at` isn't actually exposed anywhere in the user object currently, so we can remove this for now

murbanowicz commented:
I've addressed all the remarks. I had to remove type:module from package.json to fix ESLint, then change rollup config extension to mjs to get it working without it, but now it all build etc. 
I hope it's fine :)
